File: algo/MergeSort/mergeSort_basic.py
class mergeSort:
    @classmethod
    def Sort(cls, arr):
        if arr is None:
            return
        result = mergeSort().mergeSort(arr, 0, len(arr) - 1)
        for i in range(0, len(result)):
            arr[i] = result[i]
          
    def mergeSort(self, arr, start, end):
        if start >= end:
            return [arr[start]] #注意, 不是arr[start]
        middle = (start + end) // 2
        left: list = self.mergeSort(arr, start, middle)
        right: list = self.mergeSort(arr, middle + 1, end)
        return self.merge(left, right)
      
    #推倒merge, 先想著把[2, 6, 1][3, 5, 4]變成[1, 2, 3, 4, 5 ,6]
    def merge(self, arr1, arr2):
        result = []
        index1 = 0
        index2 = 0
        while(index1 < len(arr1) and index2 < len(arr2)):
            if arr1[index1] <= arr2[index2]:
                result.append(arr1[index1])
                index1 += 1
            else:
                result.append(arr2[index2])
                index2 += 1
        result.extend(arr1[index1:]) 
        result.extend(arr2[index2:])
        return result
if __name__ == "__main__":
    arr = [2, 6, 1, 3, 5, 4]
    mergeSort.Sort(arr)
    print(arr)
  
    """
    different between [arr[start]] and arr[start]

    return [arr[start]]:
    arr = [10, 20, 30]
    start = 0
    result = [arr[start]] # result [10]
    type(result)         # <class 'list'>
    return arr[start]:
    arr = [10, 20, 30]
    start = 0
    result = arr[start]  # result 10
    type(result)         # <class 'int'>
    """
    """
    what is extend method??
    add an list into another list 
    
    For example:
    my_list = ['a', 'b']
    my_tuple = ('c', 'd', 'e')
    my_list.extend(my_tuple)
    print(my_list) # output: ['a', 'b', 'c', 'd', 'e']
    """

File: algo/MergeSort/mergesort_basic.go
package main
import "fmt"

func mergesort(arr []int){
	if arr == nil{
		return
	}
	fmt.Println("init arr = ", arr)
	// mergesort_basic(arr, 0, len(arr) - 1)
	result := make([]int, len(arr))
	mergesort_advance(arr, 0, len(arr) - 1, result)
}

func mergesort_basic(arr []int, start int, end int) []int{
	if start == end{
		return []int{arr[start]}
	}
	middle := start + (end - start) / 2
	left := mergesort_basic(arr, start, middle)
	right := mergesort_basic(arr, middle + 1, end)
	return merge(left, right)
}
//推倒merge, 先想著把[2, 6, 1][3, 5, 4] 變成 [1, 2, 3, 4, 5 ,6]
func merge(arr1 []int, arr2 []int) []int {
	result := make([]int, len(arr1) + len(arr2))
	index1, index2 := 0, 0
	for index1 < len(arr1) && index2 < len(arr2){
		if arr1[index1] <= arr2[index2] {
			result[index1 + index2] = arr1[index1]
			index1++
		} else {
			result[index1 + index2] = arr2[index2]
			index2++
		}
	}
	if index1 < len(arr1){
		copy(result[index1+index2:], arr1[index1:])
	}
	if index2 < len(arr2){
		copy(result[index1+index2:], arr2[index2:])
	}
	fmt.Println("result = ", result)
	return result
}
func main(){	
	/*
	單獨測試merge
	merge([]int{2, 6, 1}, []int{3, 5, 4})
	*/
	//arr := []int{2, 6, 1, 3, 5, 4}
	arr2 := []int{7, 3, 6, 2, 0, 1, 5, 4}
	mergesort(arr2)
}

/*
代碼出處
https://leetcode.cn/leetbook/read/sort-algorithms/euivj1/

執行指令
go run ./algo/MergeSort
*/